src/utils/simplifier.js

```diff
@@ -38,11 +38,56 @@ function simplify(realm, value: Value): Value {
   switch (value.kind) {
     case "!":
       return negate(realm, value.args[0]);
+    case "||":
+    case "&&":
+      let x = simplify(realm, value.args[0]);
+      let y = simplify(realm, value.args[1]);
+      if (x instanceof AbstractValue && x.equals(y)) return x;
+      return AbstractValue.createFromLogicalOp(realm, (value.kind: any), x, y, value.expressionLocation);
+    case "==":
+    case "!=":
+    case "===":
+    case "!==":
+      return simplifyEquality(realm, value);
     default:
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (x instanceof ConcreteValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : xy) op undefined/null" to just "cond" or "!cond"
+    let [cond, xx, xy] = x.args;
+    invariant(cond instanceof AbstractValue); // otherwise the the conditional should not have been created
+    if (op === "===" || op === "!==") {
+      // if xx === undefined && xy !== undefined then cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightNotBeUndefined() && !xy.mightBeUndefined())
+        return op === "===" ? cond : negate(realm, cond);
+      // if xx !== undefined && xy === undefined then !cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightBeUndefined() && !xy.mightNotBeUndefined())
+        return op === "===" ? negate(realm, cond) : cond;
+      // if xx === null && xy !== null then cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightNotBeNull() && !xy.mightBeNull())
+        return op === "===" ? cond : negate(realm, cond);
+      // if xx !== null && xy === null then !cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightBeNull() && !xy.mightNotBeNull())
+        return op === "===" ? negate(realm, cond) : cond;
+    } else {
+      invariant(op === "==" || op === "!=");
```

Comment:
All the checks and this nested invariant seems to imply a general precondition for this function: op.kind in ["==", "!=", "===", "!=="]

Reply:
It does, but I like having the derived invariant here instead.